make empty checks outcome configurable

emattheis opened the pull request:
Per https://github.com/eclipse/microprofile-health/issues/138, this PR enables setting the desired outcome when no health checks are installed. Default behavior is preserved to return `UP`. Setting `io.smallrye.health.emptyChecksOutcome=DOWN` will override the default.

kenfinnigan commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -71,6 +73,11 @@ private static Throwable getRootCause(Throwable t) {
         return UncheckedExceptionDataStyle.ROOT_CAUSE;
     }
+
+    @Produces
```
Does this need the `@Produces`?

emattheis replied:
this seems to be a peculiarity of the CDI environment - it's unhappy without a @Default instance of the enum available

kenfinnigan replied:
That's kind of weird given it's not injected anywhere. @mkouba any thoughts on this?

emattheis replied:
yeah, I thought is was strange as well - ran into it with my previous PR and resolved with this: https://github.com/smallrye/smallrye-health/blob/master/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L70-L73 so just followed suit with this change

mkouba replied:
If it's not injected anywhere then there is no need to turn this into a bean. @emattheis what do you mean _"CDI is unhappy without a default instance"_?

emattheis replied:
@mkouba it _is_ injected by micro profile config into the `emptyChecksOutcome` field. I didn't think it would be necessary to make the enum a bean, but I had to do so with the `UncheckedExceptionDataStyle` enum or else I got a Weld exception at startup, so I followed suit with this change. I can give it a try without the producer later this morning and see what happens...

emattheis replied:
it's actually the TCK that complains at startup:

kenfinnigan commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -82,6 +89,10 @@ public static UncheckedExceptionDataStyle getDefaultUncheckedExceptionDataStyle(
+    
+    @Inject
+    @ConfigProperty(name = "io.smallrye.health.emptyChecksOutcome", defaultValue = "UP")
+    private State emptyChecksOutcome = getDefaultEmptyChecksOutcome();
```
Is it necessary to have the method that returns `State.UP` as well as setting the default config value to "UP"?

emattheis replied:
for consistency, I wanted to ensure the default value is set using the producer method in the event that the instance is created outside a micro profile config environment

kenfinnigan replied:
In what situation would you be using SmallRye Health without Config as well?

emattheis replied:
the test cases for one, but really I was just trying to follow along with the existing code base after noticing this: https://github.com/smallrye/smallrye-health/blob/master/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L75-L77

emattheis commented:
@kenfinnigan @mkouba I refactored to make the defaults more conventional - I forgot you could use static fields as producers - so we don't have the ugly static getters anymore

`@Produces` still seems unnecessary, but also not that egregious if it makes the TCK happy - I added a comment to make the usage clear

mkouba commented:
@emattheis Which TCK tests do actually fail? 

@antoinesd Do you happen to know why would any of the TCK tests try to inject `HealthCheckResponse.State`?

emattheis commented:
@mkouba @antoinesd probably easiest if you just checkout my fork and comment out the `@Produces` annotations here: https://github.com/emattheis/smallrye-health/blob/feature/emptyChecksOutcome/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L58-L62
Then, if you run `mvn clean package` you'll see the failure. Looks like the arquillian test harness fails to bootstrap properly, so it's not like a specific part of the TCK is failing.

emattheis commented:
Thanks @mkouba, that makes much more sense, now.

@kenfinnigan okay to merge?
